x/staking/keeper/cons_pubkey.go

```diff
@@ -0,0 +1,75 @@
+package keeper
+
+import (
+	"time"
+
+	"cosmossdk.io/collections"
+	"cosmossdk.io/x/staking/types"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// SetConsPubKeyRotationHistory sets the consensus key rotation of a validator into state
+func (k Keeper) SetConsPubKeyRotationHistory(
+	ctx sdk.Context, valAddr sdk.ValAddress,
+	oldPubKey, newPubKey *codectypes.Any, height uint64, fee sdk.Coin,
+) error {
+	history := types.ConsPubKeyRotationHistory{
+		OperatorAddress: valAddr.String(),
+		OldConsPubkey:   oldPubKey,
+		NewConsPubkey:   newPubKey,
+		Height:          height,
+		Fee:             fee,
+	}
+	err := k.ValidatorConsPubKeyRotationHistory.Set(ctx, collections.Join(valAddr.Bytes(), height), history)
+	if err != nil {
+		return err
+	}
+
+	if err := k.BlockConsPubKeyRotationHistory.Set(ctx, height, history); err != nil {
+		return err
+	}
+
+	ubdTime, err := k.UnbondingTime(ctx)
+	if err != nil {
+		return err
+	}
+
+	queueTime := ctx.BlockHeader().Time.Add(ubdTime)
+	if err := k.ValidatorConsensusKeyRotationRecordIndexKey.Set(ctx, collections.Join(valAddr.Bytes(), queueTime), []byte{}); err != nil {
+		return err
+	}
+
+	return k.SetConsKeyQueue(ctx, queueTime, valAddr)
+}
+
+// CheckLimitOfMaxRotationsExceed returns true if the key rotations exceed the limit, currently we are limiting one rotation for unbonding period.
+func (k Keeper) CheckLimitOfMaxRotationsExceed(ctx sdk.Context, valAddr sdk.ValAddress) (bool, error) {
+	count := 0
+	maxRotations := 1 // arbitrary value
```

Comment:
This might warrant being an on-chain param. In the meantime, let's put this into a `const`

Reply:
I think we can make this const, but for adding the param there is already a discussion.
https://github.com/cosmos/cosmos-sdk/pull/15165#discussion_r1154624632